elasticapm/base.py

```diff
@@ -188,15 +194,23 @@ def __init__(self, config=None, **inline):
             self._metrics.register(path)
         if self.config.breakdown_metrics:
             self._metrics.register("elasticapm.metrics.sets.breakdown.BreakdownMetricSet")
+        self._thread_managers["metrics"] = self._metrics
         compat.atexit_register(self.close)
         if self.config.central_config:
-            self._config_updater = IntervalTimer(
-                update_config, 1, "eapm conf updater", daemon=True, args=(self,), evaluate_function_interval=True
-            )
-            self._config_updater.start()
+            self._thread_managers["config"] = self.config
         else:
             self._config_updater = None
 
+        self.start_threads()
+
+    def start_threads(self):
+        with self._thread_starter_lock:
+            current_pid = os.getpid()
+            if self._pid != current_pid:
+                for manager in self._thread_managers.items():
+                    manager.start_thread()
+                self._pid = current_pid
```

Comment:
My brain is a bit fried so I might be wrong, but I think this has potential to introduce a race condition.

Let's say we fork after the `os.getpid()` but before we start all of our threads. This means one or more of our threads will be started twice, once now and once the next time we `start_threads()` because the pid will have changed. Fairly low risk, but since all of our code is designed to execute in a single thread, none of it is multi-thread safe.

Would it be worth attempting to stop the thread before starting it?

Reply:
As far as I understand the way forking is used by gunicorn, celery etc, it doesn't happen at any "random" moment like a thread context switch would. Instead, it happens at a defined place, e.g. after the application is imported.

The problem with prophylactically stopping threads is that at least for the `Transport`, closing it is destructive, and a new transport objects needs to be initiated. That would complicate things a lot AFAICT.

Reply:
Perfect. Thanks for the clarification, I know less than I should about their forking models and just wanted to make sure.